packages/next/src/build/webpack/plugins/slow-module-detection-plugin.ts

```diff
@@ -0,0 +1,240 @@
+import type { Compiler, Module, Compilation } from 'webpack'
+import type { CompilerNameValues } from '../../../shared/lib/constants'
+import { yellow, green, blue } from '../../../lib/picocolors'
+
+const PLUGIN_NAME = 'SlowModuleDetectionPlugin'
+
+const TreeSymbols = {
+  VERTICAL_LINE: '│  ',
+  BRANCH: '├─ ',
+} as const
+
+const PATH_TRUNCATION_LENGTH = 100
+
+// Matches node_modules paths, including pnpm-style paths
+const NODE_MODULES_PATH_PATTERN = /node_modules(?:\/\.pnpm)?\/(.*)/
+
+interface ModuleBuildTimeAnalyzerOptions {
+  compilerType: CompilerNameValues
+  slowModuleThresholdMs: number
+}
+
+const getModuleIdentifier = (module: Module): string => {
+  const debugId = module.debugId
+  return String(debugId)
+}
+
+const getModuleDisplayName = (module: Module): string | undefined => {
+  const resourcePath =
+    'resource' in module && typeof module.resource === 'string'
+      ? module.resource
+      : undefined
+
+  if (!resourcePath) {
+    return undefined
+  }
+
+  let displayPath = resourcePath.replace(process.cwd(), '.')
+
+  const nodeModulesMatch = displayPath.match(NODE_MODULES_PATH_PATTERN)
+  if (nodeModulesMatch) {
+    return nodeModulesMatch[1]
+  }
+
+  return displayPath
+}
+
+// Truncates long paths by preserving the beginning and end portions while adding ellipsis in the middle
+// Example (with PATH_TRUNCATION_LENGTH = 100):
+//   Input:  "/very/long/path/node_modules/.pnpm/some-package@1.0.0/node_modules/deep/nested/file/that/goes/on/for/quite/some/time.js"
+//   Output: "/very/long/path/node_modules/.pnpm/...ested/file/that/goes/on/for/quite/some/time.js"
+function truncatePath(path: string): string {
+  if (path.length <= PATH_TRUNCATION_LENGTH) return path
+
+  // Take the first third of the max length from the start
+  const startSegment = path.slice(0, PATH_TRUNCATION_LENGTH / 3)
+
+  // Take the last two-thirds of the max length from the end
+  const endSegment = path.slice((-PATH_TRUNCATION_LENGTH * 2) / 3)
```

Comment:
Why isn't that truncating to exactly `PATH_TRUNCATION_LENGTH` (PATH_TRUNCATION_LENGTH - 3 + "...".length)?

```suggestion
  // Take the first third of the max length from the start
  const startSegment = path.slice(0, PATH_TRUNCATION_LENGTH - 2)

  // Take the last two-thirds of the max length from the end
  const endSegment = path.slice(-(PATH_TRUNCATION_LENGTH - 1))
```